### PR TITLE
Refactor LoRA resharding, modernize NNX variable access, and streamline graph traversal

### PR DESCRIPTION
**Pull request by RexBearIU**
# Description

Refactors the LoRA mesh sharding, modernizes Flax NNX variable access, streamlines model graph traversal, and improves checkpoint restore robustness in `lora_utils.py`:

1. **Targeted In-Place LoRA Resharding (`apply_lora_to_model`)**: Replaces full-model `nnx.split`/`nnx.merge` and `logical_to_mesh_sharding` with direct, in-place resharding of `nnx.LoRAParam` instances via `jax.make_array_from_callback`. This ensures LoRA parameters are replicated across the device mesh (DDP) while frozen base weights (and `QArray` quantized weights in QLoRA) retain their original FSDP/TP sharding on device HBM, avoiding device-placement mismatches and graph recreation overhead on multi-host topologies.
2. **Streamlined Model Traversal & Caching**: Consolidates 4 separate full-graph iterations in `apply_lora_to_model` (`qwix_rngs` deletion, resharding, LoRA enablement check, submodule verification) into a single pass, and adds `@lru_cache` to `_load_lora_module_configs()`.
3. **Flax NNX Variable Modernization**: Replaces deprecated `.value` getter and setter accesses with `.get_value()`, `.set_value()`, and `.set_metadata(...)` across `lora_utils.py`.
4. **Checkpoint Restore Robustness (`restore_lora_from_path`)**: Ensures that target parameter shardings are correctly coerced into `NamedSharding(mesh, pspec)` before `jax.device_put`, and verifies that restored leaves are materialized arrays (raising an informative `ValueError` rather than leaving silent `ShapeDtypeStruct`s).
5. **Hardened Adapter Path Resolution (`load_adapter`)**: Supports both GCS paths (`gs://...`) and local filesystem directories with standard `commit_success.txt` validation.

# Tests

- Unit tests across all LoRA and SFT test suites passed (40/40):
  - `tests/unit/lora_utils_nnx_test.py` (11/11 passed)
  - `tests/post_training/unit/lora_utils_test.py` (24/24 passed)
  - `tests/post_training/unit/train_sft_test.py` (5/5 passed)
- End-to-end restore verification:
  - Custom MaxText LoRA adapter format: save, restore, merge (`apply_lora_on_base_params_nnx`), and unmerge (`unapply_lora_from_base_params_nnx`).
  - Native LoRA checkpoint format: save and in-place restore (`restore_lora_from_path`).
- End-to-end local validation:
  - `train_sft.py` with QLoRA (`lora.lora_weight_qtype=int8`) on synthetic dataset.
  - `train_sft_native.py` with standard LoRA on synthetic dataset.
  - `train_sft_native.py` with QLoRA on synthetic dataset.

# Checklist

- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: src/maxtext/utils/lora_utils.py [hunk 1]
```diff
@@ -640,47 +648,18 @@ def apply_lora_to_model(
...
-
-      lora_model = nnx.merge(graph_def, state)
+    replicated_sharding = jax.sharding.NamedSharding(mesh, jax.sharding.PartitionSpec())
+    for _, var in nnx.iter_graph(lora_model):
+      if isinstance(var, nnx.LoRAParam):
+        if hasattr(var, "set_metadata"):
+          var.set_metadata(sharding=jax.sharding.PartitionSpec(), out_sharding=None, sharding_names=None)
+        val = var.get_value() if hasattr(var, "get_value") else getattr(var, "value", None)
+        if isinstance(val, jax.Array) and val.sharding != replicated_sharding:
+          resharded_val = jax.make_array_from_callback(val.shape, replicated_sharding, lambda idx, v=val: v[idx])
```
[the remark this answers is not in the document]
Comment (RexBearIU) [reply src/maxtext/utils/lora_utils.py]: Addressed: Converted the array to a NumPy array () before passing to  to ensure pure CPU slicing.

File: src/maxtext/utils/lora_utils.py [hunk 2]
```diff
@@ -781,26 +760,37 @@ def _map_to_state(path, variable):
 
     if isinstance(curr, dict) and "value" in curr:
       matched_val = curr["value"]
+    elif hasattr(curr, "get_value"):
+      matched_val = curr.get_value()
     elif hasattr(curr, "value"):
       matched_val = getattr(curr, "value")
     else:
       matched_val = curr
```
[the remark this answers is not in the document]
Comment (RexBearIU) [reply src/maxtext/utils/lora_utils.py:765]: Addressed: Added explicit validation for  to fail fast with a descriptive ValueError if a parameter is missing from the checkpoint.

**Review by NuojCheng — APPROVED** [review]

**Review by bvandermoon — APPROVED** [review]
